File: scripts/02_kpi2_ttf.R
#scripts/02_kpi2_ttf.R
#Time-to-Fill simulado 

library(tidyverse)

# Crear carpetas de output por KPI
dir.create("outputs/tablas/02_kpi2_ttf", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/graficos/02_kpi2_ttf", recursive = TRUE, showWarnings = FALSE)

#cargar datos

df<- read.csv("data/WA_Fn-UseC_-HR-Employee-Attrition.csv")
cat("Dimensiones: ", dim(df), "\n")

#filtar los empleados que se Abandonaron (Attrition == Yes)

salidas <- df %>%
  filter(Attrition == 'Yes') %>%
  select(Department, JobRole, JobLevel, MonthlyIncome)

cat("\nTotal de bajas:", nrow(salidas), "\n" )

#simular dias para cubrir las vacantes

set.seed(42)

salidas <- salidas %>%
  mutate(
    #Dias base según el nivel de puesto (benchmarks)
    dias_base = case_when(
    JobLevel == 1 ~ 20, #Recien ingreso
    JobLevel == 2 ~ 30, #Junior
    JobLevel == 3 ~ 45, #senior/supervisor
    JobLevel == 4 ~ 65, #gerente
    TRUE ~ 85           #director nivel 5
  ),
  days_to_fill = dias_base + sample(-10:20,n(),replace =TRUE), #añade variacion aleatoria entre -10 y +20 dias
  days_to_fill = ifelse(days_to_fill < 1, 1, days_to_fill),
  costo_vacante = days_to_fill * (mean(salidas$MonthlyIncome)*12/365)
  )

timetofill_summary <- salidas %>%
  group_by(Department, JobLevel) %>%
  summarise(
    avg_ttf = mean(days_to_fill),
    median_ttf = median(days_to_fill),
    n = n()
    )


ttf_resumen <- salidas %>%
  group_by(Department,JobLevel) %>%
  summarise(
    n_vacantes = n(),
    avg_ttf = mean(days_to_fill) %>% round(2),
    median_ttf = median(days_to_fill) %>% round(2),
    min_ttf = min(days_to_fill),
    max_ttf = max(days_to_fill),
    avg_costo = mean(costo_vacante) %>% round(0),
    .groups = "drop"
  )%>%
  arrange(Department,JobLevel)
cat("\n📋 Resumen de Time-to-Fill por Departamento y Nivel:\n")
print(timetofill_summary)
cat("\n")
print(ttf_resumen)

#guardar resumen en CSV

write_csv(ttf_resumen,"outputs/tablas/02_kpi2_ttf/ttf_resumen_R.csv")
cat("\n Tabla guardada en ../outputs/tablas/02_kpi2_ttf/ttf_resumen_R.csv\n")

#Visualizacion 1: Boxplot por nivel, coloreando por depto

p1 <- ggplot(salidas, aes(x = factor(JobLevel), y = days_to_fill, fill = Department))+
  geom_boxplot(alpha = 0.7)+
  scale_fill_manual(values = c("Human Resources" = "#c0392b",
                               "Research & Development" = "#27AE60",
                               "Sales" = "#1F3C6E"))+
  labs(
    title = "Time-to-Fill por Nivel de Puesto",
    subtitle = "Distribución de días para cubrir vacantes",
    x = "Nivel de Puesto (1=entrada, 5=directivo)",
    y = "Días",
    fill = "Departamento"
  ) +
  theme_minimal() +
  theme(legend.position = "bottom")


ggsave("outputs/graficos/02_kpi2_ttf/kpi2_ttf_boxplot_R.png", p1, width = 10, height = 6, dpi = 150)
cat("✅ Boxplot guardado en outputs/graficos/kpi2_ttf_boxplot_R.png\n")

#Visualizacion 2 : Barras con promedio por departamento y nivel

p2 <- ttf_resumen %>%
  ggplot(aes(x = interaction(Department, JobLevel, sep = " - "), y = avg_ttf, fill = Department)) +
  geom_col() +
  geom_text(aes(label = avg_ttf), vjust = -0.5, size = 3) +
  scale_fill_manual(values = c("Human Resources" = "#C0392B", 
                               "Research & Development" = "#27AE60", 
                               "Sales" = "#1F3C6E")) +
  labs(
    title = "Promedio de Time-to-Fill por Departamento y Nivel",
    x = "",
    y = "Días promedio",
    fill = "Departamento"
  ) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")

ggsave("outputs/graficos/02_kpi2_ttf/kpi2_ttf_barras_R.png", p2, width = 10, height = 6, dpi = 150)
cat("✅ Boxplot guardado en outputs/graficos/kpi2_ttf_bbarras_R.png\n")


#Ultimo mostrar estadisticas globales 

costo_total <- sum(salidas$costo_vacante)
cat(sprintf("\n Costo total estimado de todas las vacantes: $%.s\n",format(costo_total, big.mark = ",",scientific = FALSE)))
cat(sprintf(" Costo promedio por vacantes: $%.s\n",format(mean(salidas$costo_vacante), big.mark = ",", scientific = FALSE)))

File: scripts/02_kpi2_ttf.py
# scripts/02_kpi2_ttf.py
# Time-to-fill simulado

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

#crear carpetas de output si se ejecuta primero esta script

os.makedirs("../outputs/tablas/02_kpi2_ttf", exist_ok= True)
os.makedirs("../outputs/graficos/02_kpi2_ttf", exist_ok= True)

#random

np.random.seed(42)

#cargar datos

df = pd.read_csv("../data/WA_Fn-UseC_-HR-Employee-Attrition.csv")
print(f"Dimensiones: {df.shape}")
print(f"Valores nulos: {df.isnull().sum().sum()}")

#filtro de colaboradores que se fueron (Attition = Yes)

salidas = df[df['Attrition']== 'Yes'][['Department', 'JobRole', 'JobLevel', 'MonthlyIncome']].copy()
print(f"\nTotal de bajas:  {len(salidas)}")

#Simular Dias para cubrir vacantes 
#días base según nivel(bechmarks)
dias_base= {1:20,2:30,3:45,4:65,5:85}
salidas['dias_base'] = salidas['JobLevel'].map(dias_base)

#añade una columna con variacion aleatoria entre -10 y +20 dias

salidas['days_to_fill'] = salidas['dias_base'] + np.random.randint(-10, 21, size=len(salidas))
salidas['days_to_fill'] = salidas['days_to_fill'].clip(lower=1) #evita negativos

#calcula el costo por vacante

salidas['costo_vacante'] = salidas['days_to_fill'] * (salidas['MonthlyIncome'].mean() * 12/365)
print(salidas)
#Resumen por departamento y nivel

ttf_resumen = (
    salidas.groupby(['Department','JobLevel'])
    .agg(
        n_vacantes = ('days_to_fill', 'count'),
        avg_ttf = ('days_to_fill', 'mean'),
        median_ttf = ('days_to_fill', 'median'),
        min_ttf = ('days_to_fill', 'min'),
        max_ttf = ('days_to_fill', 'max'),
        avg_costo = ('costo_vacante', 'mean')
    )
    .round(2)
    .reset_index()
    .sort_values(['Department','JobLevel'])
    )

print("\n📋 Resumen de Time-to-Fill por Departamento y Nivel:")
print(ttf_resumen.to_string(index=False))

#Guardar en CSV

ttf_resumen.to_csv("../outputs/tablas/02_kpi2_ttf/ttf_resumen_Py.csv", index=False)
print(f"\n Tabla guardada en ../outputs/tablas/02_kpi2_ttf/ttf_resumen_Py.csv")

# visualizaciones
# Configurar estilo
sns.set_theme(style="whitegrid")
colores_depto = {
    'Human Resources': '#C0392B',
    'Research & Development': '#27AE60',
    'Sales': '#1F3C6E'
}

# Boxplot por nivel, coloreado por departamento
plt.figure(figsize=(12, 6))
ax = sns.boxplot(
    data=salidas,
    x='JobLevel',
    y='days_to_fill',
    hue='Department',
    palette=colores_depto,
    boxprops = {'alpha':0.7}
)
plt.title('Time-to-Fill por Nivel de Puesto\nDistribución de días para cubrir vacantes', fontsize=14)
plt.xlabel('Nivel de Puesto (1=entrada, 5=directivo)')
plt.ylabel('Días')
plt.legend(title='Departamento', bbox_to_anchor=(0.5,-0.15), loc='upper center', ncol = len(colores_depto), frameon = False)
plt.tight_layout()
plt.savefig("../outputs/graficos/02_kpi2_ttf/kpi2_ttf_boxplot_Py.png", dpi=150, bbox_inches='tight')
plt.close()
print(" Boxplot guardado en outputs/graficos/02_kpi2_ttf/kpi2_ttf_boxplot_Py.png")

# gráfico de barras con promedios por departamento y nivel
# Crear etiqueta combinada
ttf_resumen = ttf_resumen.sort_values('avg_ttf', ascending=True).copy()
ttf_resumen['dept_nivel'] = (ttf_resumen['Department'].astype(str) + ' - ' + ttf_resumen['JobLevel'].astype(str))

plt.figure(figsize=(14, 6))
bars = plt.bar(
    ttf_resumen['dept_nivel'],
    ttf_resumen['avg_ttf'],
    color=[colores_depto[d] for d in ttf_resumen['Department']],
    alpha=0.8
)

# Añadir etiquetas con el valor
for bar, val in zip(bars, ttf_resumen['avg_ttf']):
    plt.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 1, f'{val:.1f}',
             ha='center', va='bottom', fontsize=9)

plt.title('Promedio de Time-to-Fill por Departamento y Nivel', fontsize=14)
plt.xlabel('')
plt.ylabel('Días promedio')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.savefig("../outputs/graficos/02_kpi2_ttf/kpi2_ttf_barras_Py.png", dpi=150, bbox_inches='tight')
plt.close()
print("Gráfico de barras guardado en ../outputs/graficos/02_kpi2_ttf/kpi2_ttf_barras_Py.png")


# Estadisticas globales en formato Moneda Check
# ============================================================
costo_total = salidas['costo_vacante'].sum()
costo_promedio = salidas['costo_vacante'].mean()

print("\n ESTADÍSTICAS GLOBALES:")
print(f"  Costo total estimado de todas las vacantes: ${costo_total:,.0f}")
print(f"  Costo promedio por vacante: ${costo_promedio:,.0f}")
print("="*75)
print("\n KPI 2 completado en Python.")